### DISCUSSION TITLE
Properly editing docker contents? How?

### DISCUSSION THREAD

**Discussion by ANK1036Official** [telekom-security/tpotce#802] [Q&A]
Sorry if this is a newbie question, but how exactly would I be able to modify the contents of the docker containers running in order to obfuscate dionaea for example by modifying the code?

I've tried to access via the docker exec command, but the filesystem is read only, and changing the config to not be read only didn't seem to work.

Thanks.

**Comment by Ganesh-403** [comment]
@ANK1036Official 
This is actually a consequence of how T-Pot is designed.

Most of the honeypot containers are intentionally started with a **read-only root filesystem** and managed entirely through Docker Compose. This improves security and ensures the containers remain reproducible. Simply changing the container at runtime (or trying to remount it read-write) is generally **not the intended workflow**. :contentReference[oaicite:0]{index=0}

If you want to **modify Dionaea's code** (or any other honeypot), the recommended approach is:

1. **Fork or clone the honeypot's source code** (e.g. Dionaea).
2. Modify the source as needed (banners, responses, protocol behavior, etc.).
3. Build your own Docker image.
4. Update the T-Pot Docker Compose configuration to use your custom image instead of the default one.

For example:

```yaml
services:
  dionaea:
    image: my-dionaea:latest
```

or, if you want Docker to build it locally:

```yaml
services:
  dionaea:
    build:
      context: ./docker/dionaea
```

T-Pot already supports customizing its Docker Compose stack, and the documentation recommends managing changes through Compose rather than editing running containers. This also means your modifications survive restarts and upgrades much better than changes made with `docker exec`. :contentReference[oaicite:1]{index=1}

If you're only trying to tweak configuration (rather than the application code), first check whether the container already exposes configuration files through bind-mounted volumes. Those are intended to be edited on the host and will persist across container recreation.

One question that would help suggest the cleanest solution:

- Are you trying to **modify Dionaea's Python/C code** to change its behavior, or are you only trying to **customize its configuration/fingerprints** (e.g. banners, responses, emulated services)? The latter is often possible without rebuilding the entire image.

If this solves your problem, feel free to mark it as the accepted answer so others can find it easily.
